src/metrics/family.rs

```diff
@@ -326,6 +327,24 @@ where
     }
 }
 
+impl<S: EncodeLabelSet, M: EncodeMetric + TypedMetric, T: Iterator<Item = (S, M)>> EncodeMetric
+    for RefCell<T>
```

Comment:
This is very hard to discover and doesn't feel Rust-y to me. Why do we need to depend on `RefCell` here?

I think it would be more idiomatic to do:

- `impl EncodeMetric for Vec<T> where T: EncodeMetric { }`
- `impl EncodeMetric for (S, M) where S: EncodeLabelSet, M: EncodeMetric { }`

Then, have the user provide a `Vec` instead of being generic over something that can be iterated. That should get rid of the `RefCell` and iterator dependency and would compose better with other usecases.

Reply:
An `Iterator::next` takes `&mut self` in order to be able to e.g. track its position within a `Vec`. Unfortunately `EncodeMetric::encode` takes `&self`.

In this pull request `RefCell` is used for interior mutability. It allows calling an `Iterator::next` taking `&mut self` from an `EncodeMetric::encode` providing `&self` only.

An alternative approach would be to change `EncodeMetric::encode` to take `&mut self`. I deemed the change and its implications not worth it, given that `Collector` is an abstraction for advanced users only.

All that said, I agree that the usage of `RefCell` is neither simple nor intuitive. @thomaseizinger do you see other alternatives to the one above?